pkg/controller/versions/util.go

```diff
@@ -138,39 +143,54 @@ func validateGateways(ctx context.Context, meta *metav1.ObjectMeta, spec *v2.Con
 			return []error{err}
 		}
 	}
+	return validateGatewaysInternal(meta, spec, smmr, allErrors)
+}
 
+func validateGatewaysInternal(meta *metav1.ObjectMeta, spec *v2.ControlPlaneSpec, smmr *v1.ServiceMeshMemberRoll, allErrors []error) []error {
 	meshNamespaces := common.GetMeshNamespaces(meta.Namespace, smmr)
 	gatewayNames := sets.NewString()
 	if spec.Gateways != nil {
+		if spec.Gateways.ClusterIngress != nil {
+			validateGatewayNamespace(clusterIngressName, &spec.Gateways.ClusterIngress.GatewayConfig, meshNamespaces, &allErrors)
+		}
+		if spec.Gateways.ClusterEgress != nil {
+			validateGatewayNamespace(clusterEgressName, &spec.Gateways.ClusterEgress.GatewayConfig, meshNamespaces, &allErrors)
+		}
 		for name, gateway := range spec.Gateways.IngressGateways {
-			validateGateway(name, &gateway.GatewayConfig, gatewayNames, meshNamespaces, meta.Namespace, &allErrors)
+			validateAdditionalGateway(name, &gateway.GatewayConfig, gatewayNames, meshNamespaces, meta.Namespace, &allErrors)
 		}
 		for name, gateway := range spec.Gateways.EgressGateways {
-			validateGateway(name, &gateway.GatewayConfig, gatewayNames, meshNamespaces, meta.Namespace, &allErrors)
+			validateAdditionalGateway(name, &gateway.GatewayConfig, gatewayNames, meshNamespaces, meta.Namespace, &allErrors)
 		}
 	}
 	return allErrors
 }
 
-func validateGateway(name string, gateway *v2.GatewayConfig, gatewayNames sets.String, meshNamespaces sets.String, meshNamespace string, allErrors *[]error) {
+func validateAdditionalGateway(name string, gateway *v2.GatewayConfig, gatewayNames sets.String, meshNamespaces sets.String, meshNamespace string, allErrors *[]error) {
+	validateGatewayNamespace(name, gateway, meshNamespaces, allErrors)
+
 	namespace := gateway.Namespace
 	if namespace == "" {
 		namespace = meshNamespace
 	}
-	if (gateway.Enabled == nil || *gateway.Enabled) && !meshNamespaces.Has(namespace) {
-		*allErrors = append(*allErrors, fmt.Errorf("namespace %q for ingress gateway %q must be part of the mesh", namespace, name))
-	}
 	namespacedName := namespacedNameString(namespace, name)
 	if gatewayNames.Has(namespacedName) {
 		*allErrors = append(*allErrors, fmt.Errorf("multiple gateways defined for %q", namespacedName))
 	} else {
 		gatewayNames.Insert(namespacedName)
 	}
+
 	if reservedGatewayNames.Has(name) {
 		*allErrors = append(*allErrors, fmt.Errorf("cannot define additional gateway named %q", name))
 	}
 }
 
+func validateGatewayNamespace(name string, gateway *v2.GatewayConfig, meshNamespaces sets.String, allErrors *[]error) {
+	if (gateway.Enabled == nil || *gateway.Enabled) && gateway.Namespace != "" && !meshNamespaces.Has(gateway.Namespace) {
+		*allErrors = append(*allErrors, fmt.Errorf("namespace %q for ingress gateway %q must be part of the mesh", gateway.Namespace, name))
```

Comment:
What about:
`namespace is not part of the mesh for the gateway %s/%s`

Reply:
This was actually an existing error message, but you're right that it could be improved.